### PR TITLE
End-to-End Testing (incomplete)

### PR DESCRIPTION
**Pull request by scottbommarito**
This PR creates end-to-end tests that test the CLI and the CodePush server simultaneously. They use a framework called nixt which allows for creating tests very easily.

(NOTE: nixt does not work on Windows unless you go into `nixt/lib/nixt/runner.js` and change the "spawn" to "exec".)

There is now a new top-level folder labeled "e2e" in which the tests reside. To run the tests, log into a CodePush server you would like to test (`code-push login`), then run `npm install` in the "e2e" folder, then `npm install` in the root, and then `gulp e2e`.

I have already configured the tests to run on VSTS (under `CodePush E2E`) and report results.

There is coverage of the following CLI commands (around 110 tests total):

- app add
- app rm
- app rename
- app ls
- deployment add
- deployment rm
- deployment rename
- deployment ls
- deployment h
- release
- patch
- promote
- rollback

List of things TODO:
- [ ] 1. configure it to run on its own without already being logged into the CLI
- [ ] 2. add coverage for more commands (access-key and collaborator tests are blocked by 1)
- [ ] 3. clean up code (split `package.ts` into multiple files!)
- [ ] 4. fix the issue with nixt on Windows

### CODE REVIEW

**Comment by scottbommarito** [comment]
failing test case is due to https://github.com/Microsoft/code-push-service/pull/473

File: e2e/input/apk.apk [hunk 1]
```diff
@@ -0,0 +1 @@
+console.log("i am a javascript file pretending to be an apk");
```
Comment (shishirx34) [inline e2e/input/apk.apk]: atleast make the casing proper. `I`

File: e2e/input/folder/folderFile.js [hunk 2]
```diff
@@ -0,0 +1 @@
+console.log("help, ive fallen into a folder and i cant get out");
```
Comment (shishirx34) [inline e2e/input/folder/folderFile.js]: *Help
*I have
*I

File: e2e/input/ipa.ipa [hunk 3]
```diff
@@ -0,0 +1 @@
+console.log("i am a javascript file pretending to be an ipa");
```
Comment (shishirx34) [inline e2e/input/ipa.ipa]: I
IPA

File: e2e/input/release1.js [hunk 4]
```diff
@@ -0,0 +1 @@
+console.log("first release");
```
Comment (shishirx34) [inline e2e/input/release1.js]: First

File: e2e/input/release2.js [hunk 5]
```diff
@@ -0,0 +1 @@
+console.log("second release");
```
Comment (shishirx34) [inline e2e/input/release2.js]: Second

File: e2e/input/release3.js [hunk 6]
```diff
@@ -0,0 +1 @@
+console.log("third release");
```
Comment (shishirx34) [inline e2e/input/release3.js]: Third

File: e2e/input/singleFile.js [hunk 7]
```diff
@@ -0,0 +1 @@
+console.log("i am a single file, all alone in this world");
```
Comment (shishirx34) [inline e2e/input/singleFile.js]: I

File: e2e/test/access-key.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import * as assert from "assert";
```
Comment (shishirx34) [inline e2e/test/access-key.ts]: Remove this file. Only add the meaningful tests and code. Do not populate the PR with unnecessary changes.

File: e2e/test/app.ts [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            done(error);
+        }
+    }
+    
+    function removeAndDone(done: MochaDone, appNameRemoved: string): (error: any) => void {
+        return (error: any) => {
+            appNamesAddedDuringTests.splice(appNamesAddedDuringTests.indexOf(appNameRemoved), 1);
+            done(error);
+        }
+    }
```
Comment (shishirx34) [inline e2e/test/app.ts]: remove `addAndDone` and `removeAndDone`
Use the frameworks functionality to add and cleanup the data that you need. Mocha provides `beforeEach` and `afterEach` hooks. Use them.

```
 beforeEach(function() {
    // runs before each test in this block
  });

  afterEach(function() {
    // runs after each test in this block
  });
```

File: e2e/test/app.ts [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                .stdout(Success.appRm(appNameAdded))
+                .run(Command.appRm(appNameAdded))
+                .on(Command.PROMPT_ARE_YOU_SURE).respond(Command.RESPONSE_ACCEPT)
+                .end(deferred.resolve);
+                
+            return deferred.promise;
+        }))
+            .then(() => {
+                done();
+            });
```
Comment (shishirx34) [inline e2e/test/app.ts]: align `.then`

File: e2e/test/collaborator.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import * as assert from "assert";
```
Comment (shishirx34) [inline e2e/test/collaborator.ts]: remove this file.

File: e2e/test/app.ts [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+                .code(201)
+                .run(Command.appAdd(newAppName))
+                .end(() => {
+                    nixt()
+                        .stderr(Error.appConflict(newAppName))
+                        .run(Command.appRename(oldAppName, newAppName))
+                        .end(addAndDone(done, newAppName));
+                });
+        });
+    });
```
Comment (shishirx34) [inline e2e/test/app.ts]: we also have `app list` `app remove` commands. Although they are the same as `ls` and `rm` respectively, we should add tests for them too.

File: e2e/test/deployment.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+                .run(Command.deploymentLs(fakeAppName))
+                .end(done);
+        });
+        
+        it("fails with invalid format", (done: MochaDone) => {
+            nixt()
+                .stderr(Error.invalidFormat())
+                .run(Command.invalidFormat(Command.deploymentLs(appName)))
+                .end(done);
+        });
```
Comment (shishirx34) [inline e2e/test/deployment.ts]: what about checking the options for deployment command? for eg: we have `deployment ls <appName> -k` which returns the deployment keys, we should verify those too.

File: e2e/test/deployment.ts [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        
+        it("fails if deployment not found", (done: MochaDone) => {
+            var fakeDeploymentName: string = "not_a_real_deployment";
+            
+            nixt()
+                .stderr(Error.deploymentNotFound(fakeDeploymentName))
+                .run(Command.deploymentRm(appName, fakeDeploymentName))
+                .on(Command.PROMPT_ARE_YOU_SURE).respond(Command.RESPONSE_ACCEPT)
+                .end(done);
+        });
```
Comment (shishirx34) [inline e2e/test/deployment.ts]: nit: we should also add a test case for an `enter` key response, which is same as sending a `Y`

File: e2e/test/deployment.ts [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+                .stdout(Success.deploymentAdd(appName, newDeploymentName))
+                .run(Command.deploymentAdd(appName, newDeploymentName))
+                .end(() => {
+                    nixt()
+                        .stderr(Error.deploymentConflict(newDeploymentName))
+                        .run(Command.deploymentRename(appName, oldDeploymentName, newDeploymentName))
+                        .end(done);
+                });
+        });
+    });
```
Comment (shishirx34) [inline e2e/test/deployment.ts]: same as app, `remove` and `list` commands coverage needs to be added.

**Comment by silhouettes** [comment]
We don't have the resources to finish this off, and we may be making very large architectural changes, so I'm closing this PR for now. We can re-open it later if needed!
